### ISSUE TITLE
Non-fatal Exception: g.b.a.k.c: HTTP 401/500

### ISSUE THREAD

**Issue by Nealsoni00** [apollographql/apollo-android#3099]
**Summary**
Firebase Crashlytics reporting the following non-fatal exception when performing a mutation: 
```
Non-fatal Exception: g.b.a.k.c: HTTP 401 
       at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:13)
       at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor$1.c(ApolloParseInterceptor.java:21)
       at com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor$executeHttpCall$1.onResponse(ApolloServerInterceptor.java:6)
       at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.java:6)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at java.lang.Thread.run(Thread.java:764)
```

```
Non-fatal Exception: g.b.a.k.c: HTTP 500 
       at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:13)
       at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor$1.c(ApolloParseInterceptor.java:21)
       at com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor$executeHttpCall$1.onResponse(ApolloServerInterceptor.java:6)
       at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.java:6)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run(Thread.java:923)
```

**Version**
Happened on Samsung, Google, Motorola, OnePlus running Android 8-11 (20 devices): 
```
    def dapolloVersion = "2.5.4"
    implementation "com.apollographql.apollo:apollo-runtime:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-coroutines-support:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-normalized-cache:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-android-support:$apolloVersion"
    implementation "com.apollographql.apollo:apollo-normalized-cache-sqlite:$apolloVersion"
```

**Description**
I can't seem to recreate it in development. I had one test device have this issue in production. 

Any advice on where to look or how to approach this issue would be greatly appreciated!

**Comment by martinbonnin** [comment]
Hi 👋  These are errors in your server: `500` is server error, `401` authorization error. 

For `401` errors, double check your authorization flow. If you're using oauth or something similar, make sure to retry the HTTP call if you receive `401`

For `500` errors, these are more generic server errors. Your best best is most likely to look at server traces and see if anything can be seen there.

**Comment by Nealsoni00** [comment]
Hey Martin! Thanks for the response. 

`500`: I've checked the server logs and retried the exact mutations resulting in 500s (copying the request body into postman) and they succeed. 

`401`: I have an interceptor that checks the response from the server to see if it's a 401 and it refreshes the access token in the case of an invalid token. I verified this code works in development  (set access token to expire in 10s and watching the interceptor refresh the token successfully)  — is there a situation where this interceptor code would not be run resulting in this error? 

Do you know what `g.b.a.k.c:` is? The 401 and 500s happen on the same device

**Comment by martinbonnin** [comment]
> retried the exact mutations resulting in 500s (copying the request body into postman) and they succeed.

Yea well you're server might be under high load or something else explaining why this doesn't happen in normal conditions

>  is there a situation where this interceptor code would not be run resulting in this error?

I would double check concurrency there. If you have 2 requests in parallel,  make sure they both get a new refresh token. It looks simple on paper but getting this 100% right is more complicated than it looks.

> `g.b.a.k.c`

 You most likely enabled R8 or proguard minification (https://developer.android.com/studio/build/shrink-code)

Overall, there's no silver bullet there. Code review and tracing will help but non-reproducible bugs are wayyy harder to hunt unfortunately.

**Comment by Nealsoni00** [comment]
> Yea well you're server might be under high load or something else explaining why this doesn't happen in normal conditions

It happened in the production testing last night when server was idle :(

> You most likely enabled R8 or proguard minification (https://developer.android.com/studio/build/shrink-code)

Ahh, minification makes sense — I'm using proguard. tx

> I would double check concurrency there. If you have 2 requests in parallel, 

In logs, the refresh access token interceptor is being called multiple times simultaneously but in the case of these 401's they are not logged to crashlytics so I would assume they are not this specific issue:

![interceptor refresh](https://user-images.githubusercontent.com/7013191/117343753-9a1c9000-ae72-11eb-8965-ec0aca2c3986.png)

After digging further, the 401s only appear if the synchronousRefreshToken fails and the chain proceeds with the original 401. I think you are right with this being due to concurrency. Do you have any suggestions on the best way to implement a mutex in the `okhttp3.Interceptor` to put the other requests on hold until a new access token can be obtained after the first failed 401? 

I currently run a synchronous refresh request and reperform request after successful refresh:
```
val newToken = synchronousRefreshToken()
if (newToken != null) {
  val modifiedRequest: Request = chain.request().newBuilder()
      .header("Authorization", "Bearer $newToken")
      .method(request.method, request.body)
      .build()
  response.close()
  response = chain.proceed(modifiedRequest)
}
```

I would ideally want something along the lines of a mutex around the above newToken section so only one interceptor refreshes the token and the others wait for the new token.

Thank you for all your help!

**Comment by martinbonnin** [comment]
Yep, mutex sounds like what you want to do here. IIRC the basic implementation that locks the whole token section could "over" refresh, meaning if two threads are waiting for the mutex, they might both refresh the token one after the other which is suboptimal but should work ultimately (Unless your backend invalidates tokens as soon as a new one is generated but it's hard to tell without knowing your backend).

**Comment by martinbonnin** [comment]
@Nealsoni00 I'm going to close this one, please leave a message if you want me to reopen it.
